toolbox/message_tools/bad_message_wizard.py

```diff
@@ -156,37 +183,202 @@ def reset_bad_message_cache():
+    queue_name_dict = {}
+    for bad_message in bad_messages:
+        for queue in bad_message['affectedQueues']:
+            if queue not in queue_name_dict:
+                queue_name_dict[queue] = 1
+            else:
+                queue_name_dict[queue] += 1
+    return queue_name_dict
+
+
+def filter_bad_messages():
+    filter_by_text = input(colored('Filter By Text: ', 'cyan'))
+
+    all_message_summaries: list = get_message_summaries()
+
+    filtered_message_summaries = filter_msgs(all_message_summaries, filter_by_text)
+
+    if not filtered_message_summaries:
+        print('No messages match your filter')
+        return
+
+    bad_message_summaries = get_bad_message_summaries(filtered_message_summaries)
+    if not bad_message_summaries:
+        show_no_bad_messages()
+        return
+
+    print(f'There are currently {len(bad_message_summaries)} bad messages matching your filter:')
+    paginate_messages(bad_message_summaries)
+
+    return
+
+
+def filter_msgs(all_message_summaries, filter_by_text):
+    filtered_msgs = []
+
+    for msg in all_message_summaries:
+        message_hash = msg['messageHash']
+        message_metadata = get_bad_message_metadata(message_hash)
+
+        for (k, v) in message_metadata[0].items():
+            for item in v.values():
+                if re.search(filter_by_text.lower(), str(item).lower()):
+                    if msg not in filtered_msgs:
+                        filtered_msgs.append(msg)
+
+    return filtered_msgs
+
+
 def show_bad_message_list():
     all_message_summaries: list = get_message_summaries()
-    bad_message_summaries = [summary for summary in all_message_summaries if not summary['quarantined']]
+
+    bad_message_summaries = get_bad_message_summaries(all_message_summaries)
     if not bad_message_summaries:
         show_no_bad_messages()
         return
-    bad_message_summaries.sort(key=lambda message: message['firstSeen'])
+
+    bad_message_summaries = sort_bad_messages(bad_message_summaries)
+
+    if not bad_message_summaries:
+        return
+
     print('')
     print(f'There are currently {len(bad_message_summaries)} bad messages:')
-    if len(bad_message_summaries) > 20:
-        print('Showing only the oldest 20')
-        bad_message_summaries = bad_message_summaries[:20]
 
-    pretty_print_bad_message_summaries(bad_message_summaries)
+    if len(bad_message_summaries) > ITEMS_PER_PAGE:
+        paginate_messages(bad_message_summaries)
+    else:
+        display_messages(bad_message_summaries, 0)
+
+
+def get_bad_message_summaries(all_message_summaries):
+    bad_message_summaries = [summary for summary in all_message_summaries if not summary['quarantined']]
+    return remove_milliseconds_from_firstseen(bad_message_summaries)
+
+
+def remove_milliseconds_from_firstseen(bad_message_summaries):
+    for msg in bad_message_summaries:
+        msg['firstSeen'] = msg['firstSeen'][0:len(msg['firstSeen']) - MILLISECOND_STR_LEN]
```

Comment:
`msg['firstSeen'] = rfc3339.parse_datetime(msg['firstSeen']).strftime("%Y-%m-%dT%H:%M:%S")` will work if you wanted to shorten and get rid of an extra constant